equinox/nn/attention.py

```diff
@@ -11,6 +11,183 @@
 from .linear import Linear
 
 
+def dot_product_attention_weights(
+    query: Array,
+    key_: Array,
+    bias: Optional[Array] = None,
+    mask: Optional[Array] = None,
+    dropout_p: float = 0.0,
+    *,
+    dropout: Optional[Dropout] = None,
+    key: Optional["jax.random.PRNGKey"] = None,
+    inference: Optional[bool] = None,
+    deterministic: Optional[bool] = None,
+):
+
+    r"""Computes dot-product attention weights given query and key.
+    Used by :func:`dot_product_attention`, which is what you'll most likely use.
+    But if you want access to the attention weights for introspection, then
+    you can directly call this function and call einsum yourself.
+    This also supports multi-query attention (https://arxiv.org/pdf/1911.02150).
+    **Arguments:**
+
+        - `query`: Query vectors. Should be a JAX array of shape
+            `(batch ... query_seq_length, num_heads, qk_size)`.
+        - `key_`: Key vectors. Should be a JAX array of shape
+            `(batch ... kv_seq_length, num_heads, qk_size)`.
+        - `mask`: Optional mask preventing attention to certain positions. Should be a
+            JAX array of shape `(num_heads, query_seq_length, kv_seq_length)`.
+        - `dropout_p`: Dropout probability on attention weights.
+        - `dropout`: Already initialized Dropout module. Unused if `dropout_p != 0.`
+          (Keyword only argument).
+        - `key`: A `jax.random.PRNGKey` used for dropout. Unused if `dropout_p = 0.`.
+            (Keyword only argument.)
+        - `inference`: As [`equinox.nn.Dropout.__call__`][]. (Keyword only
+            argument.)
+        - `deterministic`: (Deprecated in favour of `inference`.)
+
+        **Returns:**
+
+        A JAX array of shape `(batch..., num_heads, query_seq_length, kv_seq_length)`.
```

Comment:
The pattern in Equinox so far has been to not support arbitrary prepending dimensions, and let the end user `jax.vmap` if appropriate. Can this function be changed to match this convention? (i.e. have no batch dimension at all)